onnx/defs/data_type_utils.h

```diff
@@ -11,58 +11,95 @@
 #include "onnx/onnx_pb.h"
 
 namespace onnx {
+
+// ONNX domain.
+const char* const ONNX_DOMAIN = "";
+
 // String pointer as unique TypeProto identifier.
 typedef const std::string* DataType;
 
 namespace Utils {
-
 // Data type utility, which maintains a global type string to TypeProto map.
 // DataType (string pointer) is used as unique data type identifier for
 // efficiency.
-//
-// Grammar for data type string:
-// <type> ::= <data_type> |
-//            tensor(<data_type>) |
-//            seq(<type>) |
-//            map(<data_type>, <type>)
-// <data_type> :: = float | int32 | string | bool | uint8
-//                | int8 | uint16 | int16 | int64 | float16 | double
-//
-// NOTE: <type> ::= <data_type> means the data is scalar (zero dimension).
-//
-// Example: float, tensor(float), etc.
-//
 class DataTypeUtils {
  public:
+  
+  static void Register(DataType type_key, const TypeProto& type_proto) {
+      static std::mutex mutex;
+      std::lock_guard<std::mutex> lock(mutex);
+      if (GetTypeStrToProtoMap().find(*type_key) == GetTypeStrToProtoMap().end()) {
+          GetTypeStrToProtoMap()[*type_key] = type_proto;
+      }
+      else {
+          // One type is prevented from being registered multiple times
+          // from different domain intentionally.
+          assert(false);
+      }
+  }
+
   static DataType ToType(const std::string& type_str);
 
   static DataType ToType(const TypeProto& type_proto);
 
   static const TypeProto& ToTypeProto(const DataType& data_type);
 
+  static std::string GetElementTypeStr(TensorProto_DataType elem_type);
+
  private:
-  static void FromString(const std::string& type_str, TypeProto& type_proto);
 
-  static void FromDataTypeString(
-      const std::string& type_str,
-      TensorProto::DataType& tensor_data_type);
+  static std::string ToString(const TypeProto& type_proto);
 
-  static std::string ToString(
-      const TypeProto& type_proto,
-      const std::string& left = "",
-      const std::string& right = "");
+  static std::unordered_map<std::string, TypeProto>& GetTypeStrToProtoMap();
 
-  static std::string ToDataTypeString(
-      const TensorProto::DataType& tensor_data_type);
+};
+} // namespace Utils
 
-  static bool IsValidDataTypeString(const std::string& type_str);
+template<int elemT>
+struct TensorType {
+    static DataType Type() {
+        static TensorType tensor_type;
```

Comment:
static vars in header files tend to blow up badly with python extensions (which get loaded with RLDT_LOCAL and thus ODR rule for templates in different shared objects breaks). Maybe do explicit template instantiation in .cc file?

Reply:
thank you! fixed.